### ISSUE TITLE
installing the latest npm with the npm module doesn't work

### ISSUE THREAD

**Issue by tkellen** [ansible/ansible-modules-extras#137]
##### Issue Type:

Bug Report
##### Component Name:

npm
##### Ansible Version:

ansible 1.7.1
##### Environment:

OSX 10.10
##### Summary:

Attempting to install the latest npm with the npm module doesn't seem to work.
##### Steps To Reproduce:

Put the following in a playbook where node is installed

``` yaml
- name: make sure the latest npm is installed
  npm: name=npm global=true
```

...or try running the following:
https://github.com/tkellen/ansible-vagrant-example/commit/903ee91e174fe4d444e248bc01d2d74b8ccf704c
##### Expected Results:

Latest version of npm is installed.
##### Actual Results:

NPM version is unchanged.

**Comment by mpdehaan** [comment]
Can you really install npm with npm if npm isn't installed?  

Serious question... I need a bit more info here about how it was installed originally.

Can you paste the full verbose output here as well as your ansible version?

(Also, if you think you know how to fix this in the module, I'd be interested...not a node guy)

Thanks!

**Comment by tkellen** [comment]
Thanks for the quick reply @mpdehaan!  The npm binary is installed by default alongside node, but is not always the most current version.  This line installs node and npm.
https://github.com/tkellen/ansible-vagrant-example/blob/master/ansible.yml#L14-L15

Once npm is installed, you can upgrade it by running `npm install -g npm`.

I don't have time to dig into this at the moment, my workaround is currently to do this:

``` yaml
- name: make sure the latest npm is installed
  command: npm install -g npm
```

**Comment by mpdehaan** [comment]
If you pass the "global=yes" flag does it take effect?

**Comment by tkellen** [comment]
:( nope

**Comment by feldoh** [comment]
This is not an issue in at least ansible 1.9.1 and later. I think I spotted your mistake though; you are missing the state=latest parameter. The default value for state is present and of course if you are using npm it is guaranteed to be present so it will always be a no-op. If you add state=latest it will always update.

Tested by running: `ansible localhost -m npm -a "name=npm global=true state=latest"`

Without the state=latest, it did nothing as you reported, when I added the state param, it updated me to the latest version.

**Comment by visay** [comment]
I have this kind of problem with ansible 1.9.2 too.

**Command**:

`ansible localhost -m npm -a "path=/path/to/my/dir/containing/package-json state=latest"`

**package.json**:

```
{
    "name": "Dummy",
    "version": "0.0.1",
    "devDependencies": {
        "grunt": "latest",
        "grunt-contrib-clean": "~0.6.0",
        "grunt-contrib-copy": "latest",
        "grunt-contrib-watch": "latest",
        "grunt-contrib-uglify": "latest",
        "grunt-contrib-compass": "latest",
        "grunt-contrib-cssmin": "~0.10.0",
        "grunt-contrib-concat": "~0.5.0"
    }
}
```

First run, everything is fine and I get `node_modules` directory, but then I update `package.json` file adding `"grunt-contrib-imagemin": "~0.4.0"` and call `ansible localhost -m npm -a "path=/path/to/my/dir/containing/package-json state=latest"` again, nothing updated. I always get

```
localhost | success >> {
    "changed": false
}
```

But if I run `npm install` shell command manually inside the directory, it works fine and I get my new plugin for imagemin.

**Comment by feldoh** [comment]
@visay I think that's a separate issue worthy of its own ticket in that it is specifically about updates not working when using a package.json. The original title of this ticket does not relate to your question so it should have its own ticket to get noticed.

**Comment by visay** [comment]
@feldoh OK, here we go #957

**Comment by moises-silva** [comment]
@tkellen I think this issue should be closed. As indicated by @feldoh, you're just missing 'state=latest' and the default is 'present', so it's working as expected.

**Comment by geerlingguy** [comment]
See potentially-related issue (if the package isn't present already, `state=latest` doesn't install it currently): https://github.com/ansible/ansible-modules-extras/issues/1375

**Comment by moises-silva** [comment]
@geerlingguy Which btw should be addressed by PR #1663

**Comment by ansibot** [comment]
@chrishoffman, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@shane-walker, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pullrequests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pullrequest to the ansible/ansible repo.

**Comment by ansibot** [comment]
This issue was migrated to https://github.com/ansible/ansible/issues/29192
